### PR TITLE
[camera_web] Support for camera stream on web

### PR DESCRIPTION
**Pull request by TecHaxter**
This PR aims to provide support for strartImageStream and stopImageStream on Web.

[#92460](https://github.com/flutter/flutter/issues/92460)

Based on [#6944](https://github.com/flutter/plugins/pull/6944) from the archived [plugins](https://github.com/flutter/plugins) repository

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

POV: my first PR on a public repo

### CODE REVIEW

**Comment by TecHaxter** [comment]
@ditman @bparrishMines Can you help me how to resolve these Checks?

**Comment by bparrishMines** [comment]
You should be able to click on details for the failing issue. It looks like this is the cause in the integration test

```
The web-server device requires the Dart Debug Chrome extension for debugging. Consider using the Chrome or Edge devices for an improved development workflow.
Failure Details:
Failure in method: bytes is a multiple of 4
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following LateError was thrown running a test:
LateInitializationError: Field 'videoElement' has already been
initialized.

When the exception was thrown, this was the stack:
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 296:3  throw_
packages/camera_web/src/camera_service.dart.js 828:19                        set videoElement
packages/camera_web/src/camera_service.dart.js 859:27                        initialize
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50           <fn>
dart-sdk/lib/async/zone.dart 1407:47                                         _rootRunUnary
dart-sdk/lib/async/zone.dart 1308:19                                         runUnary
dart-sdk/lib/async/future_impl.dart 163:18                                   handleValue
dart-sdk/lib/async/future_impl.dart 847:44                                   handleValueCallback
dart-sdk/lib/async/future_impl.dart 876:13                                   _propagateToListeners
dart-sdk/lib/async/future_impl.dart 652:5                                    [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 722:7                                    <fn>
dart-sdk/lib/async/zone.dart 1399:13                                         _rootRun
dart-sdk/lib/async/zone.dart 1301:19                                         run
dart-sdk/lib/async/zone.dart 1209:7                                          runGuarded
dart-sdk/lib/async/zone.dart 1249:23                                         callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                             _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                              _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 181:7           <fn>

The test description was:
  bytes is a multiple of 4
═════════════════════════════════════════════════════════════════

end of failure 1
```

**Comment by TecHaxter** [comment]
@bparrishMines All the checks related to the files that were changed are passed, But some checks related to camera_android federated plugin are failing.
There are no changes made in this PR for camera_android.

**Linux_android android_device_tests_shard_1 master**
```
Test failure for example/integration_test/camera_test.dart after 2 attempts

[packages/camera/camera completed in 7m 0s]


The following packages had errors:
  packages/camera/camera:
    example/integration_test/camera_test.dart failed tests
See above for full details.
```

Kindly review and let me know if any changes are required, Thanks :)

**Comment by TecHaxter** [comment]
@bparrishMines @ditman The PR is ready for review and merge, Thanks :)

**Comment by TecHaxter** [comment]
@matanlurey @camsim99 
Can you review and identify what's wrong in **android_device_tests_shard_1** check, Thanks :)
```
* What went wrong:
A problem occurred configuring project ':flutter_plugin_android_lifecycle'.
> Could not resolve all files for configuration ':flutter_plugin_android_lifecycle:classpath'.
   > Could not download opencensus-contrib-grpc-metrics-0.21.0.jar (io.opencensus:opencensus-contrib-grpc-metrics:0.21.0)
      > Could not get resource 'https://repo.maven.apache.org/maven2/io/opencensus/opencensus-contrib-grpc-metrics/0.21.0/opencensus-contrib-grpc-metrics-0.21.0.jar'.
         > Could not GET 'https://repo.maven.apache.org/maven2/io/opencensus/opencensus-contrib-grpc-metrics/0.21.0/opencensus-contrib-grpc-metrics-0.21.0.jar'. Received status code 503 from server: Service Unavailable
> Failed to notify project evaluation listener.
   > Could not get unknown property 'android' for project ':flutter_plugin_android_lifecycle' of type org.gradle.api.Project.
   > Could not get unknown property 'android' for project ':flutter_plugin_android_lifecycle' of type org.gradle.api.Project.
```

And why android checks are failing if there are no changes related to android federated plugin in this PR?

**Comment by camsim99** [comment]
@TecHaxter Re-ran it because I've seen a similar error before and it looks like it was a flake. As to why it ran, maybe @ditman knows?

File: packages/camera/camera/CHANGELOG.md
Comment (bparrishMines) [inline packages/camera/camera/CHANGELOG.md:1]: @TecHaxter Can you revert the changes to this file? It's unrelated to this PR. It may also be why the Android test are running.

Comment (TecHaxter) [reply packages/camera/camera/CHANGELOG.md:1]: Sure, I have rebased the branch and removed that specific commit which included changes in packages/camera/camera/CHANGELOG.md

File: packages/camera/camera_web/lib/src/camera_service.dart [hunk 1]
```diff
@@ -393,4 +394,68 @@ class CameraService {
...
+    final int width = int.tryParse(widthString) ?? videoElement.videoWidth;
+    final int height = int.tryParse(heightString) ?? videoElement.videoHeight;
+
+    if (width == 0 || height == 0) {
+      throw Exception(
+        'Computed dimensions are zero: width=$width, height=$height',
+      );
+    }
+    final bool canUseOffscreenCanvas =
+        jsUtil.hasProperty(window!, 'OffscreenCanvas');
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_service.dart:437]: `canUseOffscreenCanvas` should be a constant of the plugin, it doesn't need to be checked every time that `takeFrame` is called.

File: packages/camera/camera_web/lib/src/camera_service.dart [hunk 2]
```diff
@@ -393,4 +394,68 @@ class CameraService {
...
+  ///Returns frame at a specific time using video element
+  CameraImageData takeFrame(html.VideoElement videoElement) {
+    final List<String> widthPx = videoElement.style.width.split('px');
+    final List<String> heightPx = videoElement.style.height.split('px');
+    final String widthString =
+        widthPx.isNotEmpty ? widthPx.first : '${videoElement.videoWidth}';
+    final String heightString =
+        heightPx.isNotEmpty ? heightPx.first : '${videoElement.videoHeight}';
+    final int width = int.tryParse(widthString) ?? videoElement.videoWidth;
+    final int height = int.tryParse(heightString) ?? videoElement.videoHeight;
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_service.dart:429]: Please, try using the `int videoElement.videoWidth/videoHeight` before doing all the dance of parsing strings.

Also: does this need to be determined on every captured frame? Can we do it in a resize listener of the video element and only recompute this when it changes?

Also also: the `videoWidth/videoHeight` doesn't appear to be nullable; do we really need to do any string parsing? is the `style` version more accurate than the reported `videoWidth/Height`?

File: packages/camera/camera_web/lib/src/camera_service.dart [hunk 3]
```diff
@@ -393,4 +394,68 @@ class CameraService {
...
+        'Computed dimensions are zero: width=$width, height=$height',
+      );
+    }
+    final bool canUseOffscreenCanvas =
+        jsUtil.hasProperty(window!, 'OffscreenCanvas');
+    late html.ImageData imageData;
+    if (canUseOffscreenCanvas) {
+      final html.OffscreenCanvas canvas = html.OffscreenCanvas(width, height);
+      final html.OffscreenCanvasRenderingContext2D context =
+          canvas.getContext('2d')! as html.OffscreenCanvasRenderingContext2D;
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_service.dart:442]: In both cases: do we need to create a new canvas on each `takeFrame` call? Could we create a single canvas for the appropriate width/height and keep that, and just call the appropriate `drawImage` on it, depending on it being off or onscreen?

File: packages/camera/camera_web/lib/src/camera_service.dart [hunk 4]
```diff
@@ -393,4 +394,68 @@ class CameraService {
...
+      context.drawImage(videoElement, 0, 0, width, height);
+      imageData = context.getImageData(0, 0, width, height);
+    } else {
+      final html.CanvasElement canvas =
+          html.CanvasElement(width: width, height: height);
+      final html.CanvasRenderingContext2D context = canvas.context2D;
+      context.drawImageScaled(videoElement, 0, 0, width, height);
+      imageData = context.getImageData(0, 0, width, height);
+      canvas.remove();
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_service.dart:451]: It seems `canvas` was never added to the DOM, so `canvas.remove()` is probably unnecessary?

File: packages/camera/camera_web/lib/src/camera_service.dart [hunk 5]
```diff
@@ -393,4 +394,68 @@ class CameraService {
         return DeviceOrientation.portraitUp;
     }
   }
+
+  /// Maps a [Uint] to a [CameraImageData].
+  CameraImageData getCameraImageDataFromBytes(
```
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera_service.dart:399]: This method seems a little bit superfluous, it seems its implementation could be inlined where this is used?

**Review by ditman — COMMENTED** [review]
I've added a few comments to try to minimize calls to `takeFrame` and also try to minimize the work done in that method.

Thanks for the PR!!

File: packages/camera/camera_web/lib/src/camera_service.dart (hunk 2 above)
Comment (TecHaxter) [reply packages/camera/camera_web/lib/src/camera_service.dart:429]: I checked, the string interpolation is not necessary. Removing it in next commits.
`int videoElement.videoWidth/videoHeight` is a better approach.

File: packages/camera/camera_web/lib/src/camera_service.dart (hunk 3 above)
Comment (TecHaxter) [reply packages/camera/camera_web/lib/src/camera_service.dart]:
```
  ///Used in [takeFrame] if [canUseOffscreenCanvas] is false
  html.CanvasElement? _canvasElement;

  ///Used in [takeFrame] if [canUseOffscreenCanvas] is false
  html.OffscreenCanvas? _offscreenCanvas;
```

Added these in camera_service.dart to preserve and re-use `_offscreenCanvas`/`_canvasElement` in `takeFrame()`
I'm not sure if that's a right approach to use variables in service class, let me know.

File: packages/camera/camera_web/lib/src/camera_service.dart (hunk 2 above)
Comment (TecHaxter) [reply packages/camera/camera_web/lib/src/camera_service.dart]: Apologies for that code,
We do not need to determine `videoWidth/videoHeight` on every frame, because these parameter never changes during the life-cyle of the `videoElement`. It only changes when camera is initialized.

Here's what I found in `camera_web.dart` file
```
  /// Emits an empty stream as there is no event corresponding to a change
  /// in the camera resolution on the web.
  ///
  /// In order to change the camera resolution a new camera with appropriate
  /// [CameraOptions.video] constraints has to be created and initialized.
  @override
  Stream<CameraResolutionChangedEvent> onCameraResolutionChanged(int cameraId) {
    return const Stream<CameraResolutionChangedEvent>.empty();
  }
```

Comment (TecHaxter) [reply packages/camera/camera_web/lib/src/camera_service.dart]: And something like this can be added to listen to resize changes on videoElement.

```
 final Map<int, StreamSubscription<html.Event>>
      _cameraVideoResizedSubscriptions =
      <int, StreamSubscription<html.Event>>{};
```
```
      // Add camera's resolution events to the camera events stream.
      // The onResize stream fires when videoElement gets resized
      _cameraVideoResizedSubscriptions[cameraId] =
          camera.videoElement.onResize.listen((html.Event _) {
        cameraEventStreamController.add(
          CameraResolutionChangedEvent(
            cameraId,
            camera.videoElement.width.toDouble(),
            camera.videoElement.height.toDouble(),
          ),
        );
      });
```
```
  @override
  Stream<CameraResolutionChangedEvent> onCameraResolutionChanged(int cameraId) {
    return _cameraEvents(cameraId).whereType<CameraResolutionChangedEvent>();
  }
```

But it will be non functional due to the reason mentioned in the above comment's code!

**Comment by stuartmorgan** [comment]
@ditman It looks like this ready for another review.

**Comment by kevmoo** [comment]
Are we unblocked here @ditman ?

**Comment by stuartmorgan** [comment]
From triage: Ping @ditman on this review

**Comment by Piinks** [comment]
(PR triage): @ditman can you give this another look? Apologies @TecHaxter, thank you for your patience. It looks like there are some merge conflicts now, can you take a look? Thank you!

**Comment by ditman** [comment]
This is next in my review queue, will get to it on friday. I can deal with the conflicts (if @TecHaxter allows contributors to modify this PR)

**Comment by ditman** [comment]
@TecHaxter I'm going to rebase and foce push this branch to remove all the unneeded commits/participants.

**Comment by ditman** [comment]
This PR is super outdated, it never made it through the `dart:html` migration. Unfortunately, @TecHaxter didn't follow the tree hygiene docs and created this PR on the `main` branch of their fork, so I won't be able to fix this PR.

What I'm going to do, is try to cherry-pick @TecHaxter's commits on a new branch off of `main`, and see how far I can get:

```console
dit@dit:/work/flutter/packages/packages/camera/camera_web$ git log --author=mr.khushalrao@gmail.com --no-merges --format=oneline --reverse
822c5bfe4e6663d70a251ecef2d015ce19b57f52 [camera/camera_web] Supporting camera stream on web
88dce1d8282139de2b788fe90b90ae09f397cec0 [camera_web] Version and Changelog updated
bcea8122ffc0dd729aadeb3c51f444cb8c4ad7f2 [camera_web] - integration test updated: mock videoElement initialized before expecting mock response
c776141e4b3e608c67bd7e4da3340e899f10b8c5 [camera_web] camera service takeFrame with better exception handelling
6d73310b0d4cd96d90639beb592a099ccd4efdc4 [camera_web] tested takeFrame camera service function
7244e7f5c33895cdc9fd4d40ffbcd80d76585600 [camera_web] tested cameraFrameStream
9ebbddbfeba60cb0d87cd020858861caa791cbca [camera_web] platform version updated according to guidelines
01594c48dbb3ea5f41ef7941ea9968cc4e2be51a style: removed late from late final_cameraFrameStreamController
7d83693b94e3592b22052456c21f32cfdc229f8e refactor: cameraFrameStream handles looping of animation frames
b3e8bc5cd4e4f2654dfc6aecdbd37fe46881bf18 refactor: saperated _triggerAnimationFramesLoop from cameraFrameStream, and used window.animationFrames
327dcdf9861fac4431f2d518649afe84f7ace97e chore: can use off screen canvas constant
ef44e13f6d5dfd563cf152e047b7b61789d8c01f chore: removed unnecessary string interpolation from videoElement style height and width in takeFrame
6c607056d10c5880da6976c8e1879e0f6afa3158 chore: reusing the OffscreenCanvas or CanvasElement depending on canUseOffScreenCanvas
7eceff29dffb69011011069b5de4b56e46f1b286 refactor: removed getCameraImageDataFromBytes to be used in-line
db792f415aa6847c3d2f18763feea7a7156b327b fix: setup hasPropertyOffScreenCanvas in tests
6a8c5db4a54c2c764ffe3c27925ddf18231e297f chore: version and changelog updated
4db1863e6b8560afd0950a634f0fc5a9091e33f9 docs: removed "Streaming of frames" from Missing implementation
56e5abbc8f44984e0ba8605816d5cdbb64cd22cb (HEAD, TecHaxter/main) Revert "docs: removed "Streaming of frames" from Missing implementation"
```

**Comment by TecHaxter** [comment]
> This is next in my review queue, will get to it on friday. I can deal with the conflicts (if @TecHaxter allows contributors to modify this PR)

@ditman The following checkbox is marked true on this PR:
- [x] Allow edits by maintainers

Let me know if I requires to do anything with the PR. 
Thanks :)

**Comment by TecHaxter** [comment]
Hi @ditman,

I've resolved the merge conflicts, and the latest commits replace the `dart:html` imports with `package:web`.

Since `window.animationFrame` is unavailable in `package:web`, the implementation now utilizes `window.requestAnimationFrame` instead.

Additionally, the `_triggerAnimationFramesLoop` function has been updated to be non-blocking and non-recursive (in a technical sense). Currently, a static FPS value of 30 is being used, as [requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame) can execute faster on high-refresh-rate screens, leading to screen freezing if the FPS is not controlled.

We can remove this static FPS value later once FPS is introduced in `CameraImageStreamOptions` (the platform interface class).

**Comment by TecHaxter** [comment]
Sorry for the trouble, @ditman . I've created a new PR [#7950](https://github.com/flutter/packages/pull/7950) by cherry picking required commits from this PR. Should we close this PR and keep the new one?, or is there any other way to continue with this PR? Let me know how you'd like to proceed. Thanks!

**Comment by TecHaxter** [comment]
@ditman I have just updated the rules on my fork repository `main` branch, it is open to all contributors (force push allowed)
I appreciate your help, thanks :)

**Comment by ditman** [comment]
> Should we close this PR and keep the new one?, or is there any other way to continue with this PR? Let me know how you'd like to proceed.

@TecHaxter thanks! You saved me a lot of trouble, haha!

Lets' close this PR, and continue on the new one. I'll try to review/change stuff over there!

> I've updated the rules on my fork repository main branch, it is open to all contributors (force push allowed)

Now that you're working on a separate branch, you can keep your `main` branch protected (in fact, I rarely push to the `main` of my fork, I always use the ["Sync fork" Github button](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-branch-from-the-web-ui) to keep it updated!

**Comment by ditman** [comment]
Moved to: https://github.com/flutter/packages/pull/7950
